resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/AliasManager.java

```diff
@@ -624,27 +625,41 @@ public void updateAliases(IResource resource, IFileStore location, int depth, IP
 	/**
 	 * Process any structural changes that have occurred since the last alias
 	 * request.
+	 *
+	 * @return <code>true</code> if some structural changes where processed,
+	 *         <code>false</code> if not structural changes were detected
 	 */
-	private void updateStructureChanges() {
+	private boolean checkStructuralChanges() {
+		try {
+			updating.acquire();
+		} catch (InterruptedException e) {
+			ResourcesPlugin.getPlugin().getLog()
+					.log(new Status(IStatus.ERROR, ResourcesPlugin.PI_RESOURCES, e.getMessage(), e));
+			return false;
+		}
 		boolean hadChanges = false;
-		if (changedProjects) {
+		while (changedProjects.get()) {
 			//if a project is added or removed, just recompute the whole world
-			changedProjects = false;
+			changedProjects.set(false);
 			hadChanges = true;
 			buildLocationsMap();
-		} else {
-			//incrementally update location map for changed links
-			for (IResource resource : changedLinks) {
+		}
+		// incrementally update location map for changed links
+		while (!changedLinks.isEmpty()) {
+			Collection<IResource> changedLinksSnapshots = new HashSet<>(changedLinks);
+			changedLinks.removeAll(changedLinksSnapshots);
+			for (IResource resource : changedLinksSnapshots) {
 				hadChanges = true;
 				if (!resource.isAccessible())
 					continue;
 				if (resource.isLinked())
 					addToLocationsMap(resource, ((Resource) resource).getStore());
 			}
 		}
-		changedLinks.clear();
-		if (hadChanges)
+		if (hadChanges) {
 			buildAliasedProjectsSet();
-		changedProjects = false;
+		}
+		updating.release();
```

Comment:
Just a note: If **anything** goes wrong, e.g. a runtime exception is thrown in any of the action before the semaphore will never be released at all it also seems that a semaphore is used here without real need for such and a simple synchronized and/or lock will be enough.